## 1. Import Libraries

Libraries for data processing and modeling.

- **pandas / numpy** → data handling and numerical operations  
- **train_test_split** → splitting data into training and testing sets  
- **LogisticRegression** → baseline model for fraud prediction  
- **classification_report / confusion_matrix** → evaluating model performance  

In [1]:
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report, confusion_matrix

## 2. Load Data

Load the transaction dataset into a DataFrame.

The dataset contains transaction type, amount, account balances, and fraud labels.

In [6]:
df_raw = pd.read_csv("../data/PS_20174392719_1491204439457_log.csv")

## 3. Data Preparation

Select only the relevant columns for modeling.

This helps simplify the dataset and focus on key features related to fraud detection.

In [9]:
df_selected = df_raw[['type', 'amount', 'oldbalanceOrg', 'newbalanceOrig', 'isFraud']]

## 4. Feature Encoding

Convert the transaction type column into numerical format using one-hot encoding.

This allows the model to understand different transaction categories.

In [10]:
df_encoded = pd.get_dummies(df_selected, columns=['type'], drop_first=True)

## 5. Train/Test Split

Split the dataset into training and testing sets.

The model learns from the training data and is evaluated on unseen test data.

In [11]:
X = df_encoded.drop('isFraud', axis=1)
y = df_encoded['isFraud']

## 6. Model Training

Train a Logistic Regression model as a baseline.

This model is simple, fast, and commonly used for binary classification problems like fraud detection.

In [12]:
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

## 7. Model Evaluation

Evaluate the model using confusion matrix and classification report.

Since fraud cases are rare, recall and precision are more important than accuracy.

In [13]:
model = LogisticRegression(max_iter=1000)
model.fit(X_train, y_train)

LogisticRegression(max_iter=1000)

## 8. Prediction

Use the trained model to predict whether each transaction is fraud (1) or not (0).

These predictions will be used to evaluate model performance on unseen data.

In [15]:
y_pred = model.predict(X_test)

## 9. Model Evaluation

Evaluate the model using:

- Confusion Matrix → shows correct and incorrect predictions  
- Precision → how many predicted fraud cases are correct  
- Recall → how many fraud cases are detected  
- F1-score → balance between precision and recall  

Since fraud cases are rare, recall and precision are more important than accuracy.

In [16]:
print(confusion_matrix(y_test, y_pred))
print(classification_report(y_test, y_pred))

[[1270270     634]
 [     33    1587]]
              precision    recall  f1-score   support

           0       1.00      1.00      1.00   1270904
           1       0.71      0.98      0.83      1620

    accuracy                           1.00   1272524
   macro avg       0.86      0.99      0.91   1272524
weighted avg       1.00      1.00      1.00   1272524



## Business Interpretation

- The model detects most fraud cases (~98%)
- Very few fraud transactions are missed
- Some normal transactions are flagged as fraud

### What this means

- Fraud detection is strong and reliable  
- Risk of missing fraud is low  
- Some transactions may require manual review  

### Conclusion

The model is effective and can be used as a first step in fraud detection.